File: clients/js/src/load-feature-specs.ts
import fs from "fs";
import path from "path";
import yaml from "yaml";

export interface TestCases {
  [testCase: string]: {
    input: unknown;
  }
}

export interface Spec {
  required: boolean;
  cases: TestCases;
}

export interface FeatureSpecs {
  [spec: string]: Spec;
}

export function loadFeatureSpecs(directory: string): FeatureSpecs {
  const featureSpecs: FeatureSpecs = { };

  const specFiles = fs.readdirSync(
    directory
  );

  for (const specFile of specFiles) {
    const specYaml = fs.readFileSync(
      path.join(directory, specFile),
      "utf-8"
    );

    const spec = yaml.parse(specYaml);

    if (!spec || typeof spec !== "object") {
      throw Error(`Failed to load feature-spec ${specFile}, must be an object.`);
    }

    const required = spec.required;

    if (typeof required !== "boolean") {
      throw Error(`Failed to load feature-spec ${specFile}, must have property 'required'.`);
    }

    const cases = spec.cases;

    if (typeof cases !== "object") {
      throw Error(`Failed to load feature-spec ${specFile}, must have property 'cases'.`);
    }

    for (const testCase of Object.keys(cases)) {
      if (!cases[testCase].input) {
        throw Error(
          `Failed to load feature spec test case ${specFile}.cases.${testCase}, missing 'input' property`
        );
      }
    }

    const specName = specFile.replace(
      path.extname(specFile),
      ""
    );

    featureSpecs[specName] = spec;
  }

  return featureSpecs;
}

File: clients/py/src/load_feature_specs.py
import os
import yaml

from typing import Any, Dict

from pydantic import BaseModel


class TestCases(BaseModel):
    input: Any


class Spec(BaseModel):
    required: bool
    cases: Dict[str, TestCases]


def load_feature_specs(directory: str) -> Dict[str, Spec]:
    feature_specs: Dict[str, Spec] = {}

    spec_files = os.listdir(directory)

    for spec_file in spec_files:
        with open(os.path.join(directory, spec_file), "r") as f:
            spec_yaml = f.read()
            spec = Spec.validate(yaml.load(spec_yaml, Loader=yaml.SafeLoader))
            spec_name = os.path.splitext(spec_file)[0]
            feature_specs[spec_name] = spec

    return feature_specs
